Grace hash join supports right/full join

lgbo-ustc opened the pull request:
<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/run_check.py
-->
### Changelog category (leave one):
- New Feature



### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Support full/right join by using grace hash join algorithm.

fixes: #48546

### Documentation entry for user-facing changes

- [x] Documentation is written (mandatory for new features)

<!---
Directly edit documentation source files in the "docs" folder with the same pull-request as code changes

or

Add a user-readable short description of the changes that should be added to docs.clickhouse.com below.

At a minimum, the following information should be added (but add more as needed).
- Motivation: Why is this function, table engine, etc. useful to ClickHouse users?

- Parameters: If the feature being added takes arguments, options or is influenced by settings, please list them below with a brief explanation.

- Example use: A query or command.
-->


> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

lgbo-ustc committed:
grace hash join supports right/full join

robot-clickhouse commented:
<!-- automatic status comment for PR #49483 from bigo-sg/ClickHouse:grace_hash_full_join -->
This is an automated comment for commit 1b1906c9c62120fe095c07d787ee7ea8ec0a0382 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/49483/1b1906c9c62120fe095c07d787ee7ea8ec0a0382/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🔴 failure</td></tr>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>🟡 pending</td></tr>
</table>

lgbo-ustc committed:
add comment

vdimir commented on tests/queries/0_stateless/02724_grace_hash_join.sql:
```diff
@@ -0,0 +1,44 @@
+set join_algorithm = 'grace_hash';
```
We can also reuse existing tests. Some tests are testing several algorithms and for grace_join there's check that it doesn't support `RIGHT` and `LEFT`. We can remove it and the result should be the same as for another algorithms.

https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/01721_join_implicit_cast_long.sql.j2#L12-L14

https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/02273_full_sort_join.sql.j2#L30-L32

https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/02274_full_sort_join_nodistinct.sql.j2#L18-L20

https://github.com/ClickHouse/ClickHouse/blob/master/tests/queries/0_stateless/02275_full_sort_join_long.sql.j2#L25-L29

lgbo-ustc replied:
thinks very much for this.

lgbo-ustc committed:
update

1. fixed the memoery overflow problem when  handle all delayed buckets parallely
2. resue exists tests

lgbo-ustc committed:
update tests

lgbo-ustc committed:
triger ci

lgbo-ustc committed:
update

lgbo-ustc committed:
add locks for getNonJoinedBlocks

lgbo-ustc committed:
update

lgbo-ustc committed:
update

lgbo-ustc committed:
fixed: return invalid mismatch rows on full/right join

lgbo-ustc committed:
ensure only the last processor could access non-joined blocks

lgbo-ustc committed:
fixed: a in-memory bucket contains rows of other buckets

lgbo-ustc commented:
There are still two bugs to be fixed which may be some compatibility issues.

lgbo-ustc committed:
update tests for using clause

lgbo-ustc commented:
@vdimir  Is there any way to skip the test for `grace_hash` in the `flaky` tests. It's too slow with small `max_block_size`.

vdimir commented on src/Interpreters/GraceHashJoin.cpp:
```diff
@@ -555,16 +566,6 @@ IBlocksStreamPtr GraceHashJoin::getDelayedBlocks()
 
-    if (hash_join)
-    {
-        auto right_blocks = hash_join->releaseJoinedBlocks(/* restructure */ false);
```
We can have rows that belong to different bucket.
For example:
we have right keys [0,2,4] in 0-bucket and [1,3,5] in the 1-bucket and we are reading the right table.
On addJoinedBlock we add keys [6,7] have overflow and do rehash.
Now we have buckets with right keys:
right keys | bucket
--- | ---
[0,2,4] | 0
[1,3,5] | 1
[6]     | 2
[7]     | 3

Then we read left table and add rows to buckets also:
left | right | bucket
--- | --- | ---
[0] | [0,2,4] | 0
[1] | [1,3,5] | 1
[2] | [6]     | 2
[3] | [7]     | 3

Then we go to delayed stage. We read bucket 1 and process keys [1,5] wich belong to it, but row with the key [3] shold go to bucket 3, because we have corresponding left row there that was added after rehash from 2 to 4 buckets. 

Will we handle it in new code in `addJoinedBlockImpl`? Could you explain?

lgbo-ustc replied:
There a some rehash operation to avoid this case happen, and the buckets are hadled one after one.

When inserting [6,7] trigger rehash,  bucket[0] will releases all blocks and make a rehash them by the new bucket size, other buckets remain unchanged. So the contents of buckets now are
|  right keys   | bucket  |
|  ----  | ----  |
| [0,4] | 0 |
| [1,3,5]  | 1 |
| [2,4] | 2 |
| [7] | 3 |

When come to  join left blocks(say we have keys [0,1,2,3] in left table) but not in the delayed stage, buckets' size will not be changed, so bucket[0] will be find. What we have after bucket[0]  is
|  left keys   |  right keys   | bucket  |
|  ----  |  ----  | ----  |
| [1] | [1,3,5]  | 1 |
| [2] | [2,4] | 2 |
|[3] | [7] | 3 |

bucket[0]'s right side contains some rows not belong to it, But during loading the right blocks from bucket[0], there also a `scatter` operation. It will deliver key [3] to the bucket[3], now we  have
|  left keys   |  right keys   | bucket  |
|  ----  |  ----  | ----  |
| [1] | [1,5]  | 1 |
| [2] | [2,4] | 2 |
|[3] | [3,7] | 3 |

Then take the  probe stage for bucket[1] left side. Also blocks loading from file will also be `scatter` again, if the bucket[1] left side contains keys of other buckets, they will also be delivered to the new buckets.

lgbo-ustc commented on tests/queries/0_stateless/02274_full_sort_join_nodistinct.sql.j2:
```diff
@@ -47,9 +46,10 @@ SELECT t1.key, t2.key, length(t1.s), t2.s FROM {{ t1 }} AS t1 {{ kind }} RIGHT J
 SELECT '{{ t1 }} ALL FULL JOIN {{ t2 }} | bs = {{ block_size }}';
 SELECT t1.key, t2.key, length(t1.s), length(t2.s) FROM {{ t1 }} AS t1 {{ kind }} FULL JOIN {{ t2 }} AS t2 ON t1.key == t2.key ORDER BY t1.key, t2.key, length(t1.s), length(t2.s); {{ is_implemented(join_algorithm) }}
 
+{% if join_algorithm == 'full_sorting_merge' or t2 != 'tn2' -%}
```
Yes, but I could find a easy way to solve this

lgbo-ustc commented on src/Interpreters/GraceHashJoin.cpp:
```diff
@@ -471,17 +472,24 @@ bool GraceHashJoin::alwaysReturnsEmptySet() const
-    /// We do no support returning non joined blocks here.
-    /// TODO: They _should_ be reported by getDelayedBlocks instead
-    return nullptr;
+    return hash_join->getNonJoinedBlocks(left_sample_block_, result_sample_block_, max_block_size_);
```
Sorry, the comment may be ambiguous. Here I mean the latest bucket. Eache bucket are handle by following steps
1. join left side with the hash table
2. read non-joined rows from the hash table

vdimir replied:
@lgbo-ustc so, because scatter is also performed during loading, the resulting hash_join will have only corresponding rows according to the current number of buckets, is it correct?

lgbo-ustc committed:
update

vdimir commented:
>  Is there any way to skip the test for grace_hash in the flaky tests. It's too slow with small max_block_size and hash join size limit

@lgbo-ustc lets make this test smaller, e.g. https://github.com/ClickHouse/ClickHouse/commit/50d27cc218652c71e648bf5b97c9bb24ca170f06 if it will be enough. (also it commit adds line to dock to pass Docs Check)

lgbo-ustc committed:
update tests for flaky test

lgbo-ustc commented:
> > Is there any way to skip the test for grace_hash in the flaky tests. It's too slow with small max_block_size and hash join size limit
> 
> @lgbo-ustc lets make this test smaller, e.g. [50d27cc](https://github.com/ClickHouse/ClickHouse/commit/50d27cc218652c71e648bf5b97c9bb24ca170f06) (I can't push commit to PR directly) if it will be enough. (also it commit adds line to dock to pass Docs Check)

done

lgbo-ustc commented:
There is an OOM error, but I cannot find out what's wrong

lgbo-ustc commented:
> Could you rebase branch one more time please, `Mergeable Check` is blocked due to broken `Unit tests (asan) — PASSED record not found` in master

done

vdimir commented:
Can it be related? Logical error appeared due to race https://github.com/ClickHouse/ClickHouse/issues/50220
It's suspicious, but I don't see how the change could affect, though.

lgbo-ustc commented:
> @lgbo-ustc Can it be related? Logical error appeared due to race #50220 It's suspicious, but I don't see how the change could affect, though.

I‘m looking into it. But I guess the modification only affects the delayed phase and the exception happened in filling right side  phase.
